File: backend/src/modules/payments/dto/create-payment.dto.ts
import { IsString, IsNotEmpty, IsOptional, IsNumber, IsDateString, IsEnum, Min } from 'class-validator';
import { ApiProperty, ApiPropertyOptional } from '@nestjs/swagger';

export class CreatePaymentDto {
  @ApiProperty({ description: 'Invoice ID this payment is for' })
  @IsString()
  @IsNotEmpty()
  invoiceId: string;

  @ApiProperty({ description: 'Payment provider name (e.g. stripe, paystack, manual)', example: 'manual' })
  @IsString()
  @IsNotEmpty()
  provider: string;

  @ApiProperty({ description: 'Payment amount', example: 49.99 })
  @IsNumber()
  @Min(0)
  amount: number;

  @ApiProperty({ description: 'Currency', example: 'USD' })
  @IsString()
  @IsNotEmpty()
  currency: string;

  @ApiProperty({ description: 'Payment status', enum: ['PROCESSING', 'SUCCEEDED', 'FAILED', 'REFUNDED'] })
  @IsEnum(['PROCESSING', 'SUCCEEDED', 'FAILED', 'REFUNDED'] as const)
  status: 'PROCESSING' | 'SUCCEEDED' | 'FAILED' | 'REFUNDED';

  @ApiPropertyOptional({ description: 'Provider transaction ID' })
  @IsOptional()
  @IsString()
  providerTransactionId?: string;

  @ApiPropertyOptional({ description: 'Failure reason (for FAILED payments)' })
  @IsOptional()
  @IsString()
  failureReason?: string;

  @ApiPropertyOptional({ description: 'Backdate createdAt (ISO 8601). For data imports.' })
  @IsOptional()
  @IsDateString()
  createdAt?: string;
}

File: sdks/rust/src/api/types/create_payment_dto.rs
pub use crate::prelude::*;

/// Request type for API operation
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct CreatePaymentDto {
    /// Invoice ID this payment is for
    #[serde(rename = "invoiceId")]
    pub invoice_id: String,
    /// Payment provider name (e.g. stripe, paystack, manual)
    pub provider: String,
    /// Payment amount
    pub amount: f64,
    /// Currency
    pub currency: String,
    /// Payment status
    pub status: CreatePaymentDtoStatus,
    /// Provider transaction ID
    #[serde(rename = "providerTransactionId")]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub provider_transaction_id: Option<String>,
    /// Failure reason (for FAILED payments)
    #[serde(rename = "failureReason")]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub failure_reason: Option<String>,
    /// Backdate createdAt (ISO 8601). For data imports.
    #[serde(rename = "createdAt")]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub created_at: Option<String>,
}
